### PR TITLE
gh-3418 Provide Access Token to the CTR

### PR DESCRIPTION
**Pull request by ghillert**
- Provide the OAuth Access Token as a command-line argument to the CTR (if available)
- Move `OAuth2AccessTokenProvidingClientHttpRequestInterceptor` to Common Security
  - This will allow for re-use of this class in the Composed Task Runner

resolves https://github.com/spring-cloud/spring-cloud-dataflow/issues/3418

depends on:

Add support for OAuth2/OpenIDConnect token based authentication
https://github.com/spring-cloud-task-app-starters/composed-task-runner/issues/62

Add OAuth2AccessTokenProvidingClientHttpRequestInterceptor
https://github.com/spring-cloud/spring-cloud-common-security-config/issues/58

### CODE REVIEW

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/DataFlowControllerAutoConfiguration.java [hunk 1]
```diff
@@ -34,6 +34,7 @@
 import org.springframework.boot.context.properties.EnableConfigurationProperties;
 import org.springframework.boot.web.client.RestTemplateBuilder;
 import org.springframework.cloud.common.security.AuthorizationProperties;
+import org.springframework.cloud.common.security.support.OAuth2AccessTokenProvidingClientHttpRequestInterceptor;
```
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/DataFlowControllerAutoConfiguration.java]: Unused import.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultTaskExecutionService.java [hunk 2]
```diff
@@ -185,6 +195,46 @@ public long executeTask(String taskName, Map<String, String> taskDeploymentPrope
...
+		if (taskConfigurationProperties.getComposedTaskRunnerName().equals(taskExecutionInformation.getTaskDefinition().getTaskName())) {
+			boolean containsAccessToken = false;
+
+			for (String commandLineArg : commandLineArgs) {
+				if (commandLineArg.startsWith("--dataflow-server-access-token")) {
+					containsAccessToken = true;
+				}
+			}
+
+			if (!containsAccessToken) {
```
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultTaskExecutionService.java]: Probably want to add a test.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/DataFlowControllerAutoConfiguration.java (hunk 1 above)
Comment (cppwfs) [reply spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/DataFlowControllerAutoConfiguration.java]: And Line 375 needs to be removed.

Comment (ghillert) [reply spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/DataFlowControllerAutoConfiguration.java]: This is being used for communicating with Skipper.

Comment (cppwfs) [reply spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/DataFlowControllerAutoConfiguration.java]: At this point its failing to find the import and failing the compile.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultTaskExecutionService.java [hunk 3]
```diff
@@ -185,6 +195,46 @@ public long executeTask(String taskName, Map<String, String> taskDeploymentPrope
 
 		TaskExecution taskExecution = taskExecutionRepositoryService.createTaskExecution(taskName);
 
+		if (taskConfigurationProperties.getComposedTaskRunnerName().equals(taskExecutionInformation.getTaskDefinition().getTaskName())) {
```
[the remark this answers is not in the document]
Comment (ghillert) [reply spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultTaskExecutionService.java]: Should be resolved.

**Comment by cppwfs** [comment]
Project fails to build.   You can look at the travis build to see the same error I'm seeing on my side.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultTaskExecutionService.java [hunk 4]
```diff
@@ -185,6 +186,26 @@ public long executeTask(String taskName, Map<String, String> taskDeploymentPrope
 
 		TaskExecution taskExecution = taskExecutionRepositoryService.createTaskExecution(taskName);
 
+		if (StringUtils.hasText(taskExecutionInformation.getTaskDefinition().getDslText())
+				&& TaskServiceUtils.isComposedTaskDefinition(taskExecutionInformation.getTaskDefinition().getDslText())) {
+			boolean containsAccessToken = false;
+
+			for (String commandLineArg : commandLineArgs) {
+				if (commandLineArg.startsWith("--dataflow-server-access-token")) {
```
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultTaskExecutionService.java]: We need to pass this is in as a property.  It being used as a command-line arg means the token will be stored in the database.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultTaskExecutionService.java (hunk 2 above)
Comment (ghillert) [reply spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultTaskExecutionService.java]: Added tests

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultTaskExecutionService.java (hunk 4 above)
Comment (ghillert) [reply spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultTaskExecutionService.java]: Made the change

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/service/impl/DefaultTaskExecutionServiceTests.java [hunk 5]
```diff
@@ -492,6 +499,131 @@ public void executeComposedTask() {
 			assertNull(request.getDefinition().getProperties().get("globalstreamkey"));
 		}
 
+		@Test
+		@DirtiesContext
+		public void executeComposedTaskWithAccessToken() {
+			String dsl = "AAA && BBB";
+			initializeSuccessfulRegistry(appRegistry);
+
+			final OAuth2Authentication oAuth2Authentication = mock(OAuth2Authentication.class);
```
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/service/impl/DefaultTaskExecutionServiceTests.java]: The common OAuth2Authentication prep and mock setup can be moved into a method and shared across the methods.    I can do this on merge

**Comment by cppwfs** [comment]
Rebased, Polished Tests on Merge, Squashed ,Merged

**Pull request closed without merging**